### ISSUE TITLE
Serialization of a std::map<std::string, ChLinkTSDA> leads to a segmentation fault

### ISSUE THREAD

**Issue by paulotto** [projectchrono/chrono#507]
Hi everyone,

I tried to serialize a class member of type *std::map<std::string, ChLinkTSDA>* with *archive_out << CHNVP(map_);*, but unfortunately, this led to a segmentation fault.

While looking for the cause of this, I came across another issue, which I wondered if was related to the first issue. If you want to do something like this
```cpp
std::map<std::string, ChLinkTSDA> map_{
    {"l1", ChLinkTSDA()}
};
```

a 'bad_alloc' exception is thrown. I could eliminate this by adding the following line to the copy constructor of ChLinkTSDA:
```cpp
m_jacobians = other.m_jacobians ? new SpringJacobians(*other.m_jacobians) : nullptr;
```

But in the end, I couldn't solve the serialization problem. Maybe you have a solution.

Thanks and best regards,
Paul

**Comment by DanNegrut** [comment]
Thanks for the detailed report, and I am sorry it sat this long without a reply.

Both halves of what you found were real. One was fixed shortly after you filed, and the other was
never fixed and still reproduces today, so this issue stays open.

### The `bad_alloc` is fixed

Your diagnosis was correct, and so was your patch. The old copy constructor never assigned
`m_jacobians`, and never assigned `m_variables` on the path where the source had no ODE registered,
while `~ChLinkTSDA()` deletes both. A default-constructed `ChLinkTSDA()` copied into a `std::map`
therefore handed two uninitialized pointers to `delete`, which corrupts the allocator's bookkeeping and
surfaces as a failure at some later, unrelated allocation.

That was fixed in 85725a0277 ("Fix copy constructors for several ChLink classes") on 2024-07-15, three
days after you opened this, and nobody connected it back to your report. Sorry about that. The current
constructor sets `m_jacobians = nullptr` unconditionally and assigns `m_variables` on both branches.

Confirmed by rebuilding with just that constructor reverted: your snippet crashes at map construction,
and with current `main` it does not.

### The serialization failure is real and is not fixed

It no longer crashes, which is arguably worse, because it now fails silently. Serializing
`std::map<std::string, ChLinkTSDA>` today writes a 254 byte record containing only the `ChObj` base
fields. Every spring parameter and both bodies are missing, and there is no `_version_ChLinkTSDA` at
all. You get a valid-looking JSON file describing nothing useful.

The cause is not in `ChLinkTSDA`. `ChArchiveOut`'s wrappers for `std::map` and `std::unordered_map`
bind their `ChNameValue` to a temporary pair, because they declare the key as non-const while a map
iterator yields `std::pair<const Key, Value>&`. `ChNameValue` stores only a raw pointer, so the write
reads memory that has already been freed. It crashed for you in 2024 and silently truncates now, which
is the usual range of behavior for undefined behavior.

Tracked as #776, with a fix proposed in #774. It affects any polymorphic Chrono object stored by
value in either associative container, so your report was considerably more significant than its title
suggests. `std::vector` and `std::list` were never affected.

### What to do in your own code

If you still care about this, hold `std::shared_ptr<ChLinkTSDA>` rather than the object by value:

```cpp
std::map<std::string, std::shared_ptr<ChLinkTSDA>> links;
```

Verified working: it serializes a complete record and resolves the attached bodies properly. Two
independent reasons to prefer it even after #774 is merged:

1. `ChLink`'s copy constructor sets both body pointers to null, so a container of links held by value
   loses its body attachments on every insert, whatever the serializer does. Filed as #777.
2. `ChLinkTSDA` declares a copy constructor and a destructor but no copy assignment operator, so the
   implicitly generated one shallow-copies `m_variables` and `m_jacobians`. The archive read path
   inserts with `container[key] = value`, so a by-value map of links is on thin ice on the way back in
   as well.

If you are still working with this, `#777` and `#778` may also be relevant, the latter
because JSON and XML currently round trip doubles at only 6 significant digits.

Thanks again for the report, and for the diagnosis you did on it.

**Comment by paulotto** [comment]
Thanks for the explanation! That clarifies things well.

Best,
Paul
